File: frontend/src/lib/canonical-intake-state.ts
type IdentityState = { relationship: string; gender: string };
type IdentityPatch = IdentityState;

const RELATIONSHIP_FACTS: Record<string, IdentityPatch> = {
  mom: { relationship: "Mom", gender: "Female" },
  mother: { relationship: "Mom", gender: "Female" },
  "my mom": { relationship: "Mom", gender: "Female" },
  "my mother": { relationship: "Mom", gender: "Female" },
  dad: { relationship: "Dad", gender: "Male" },
  father: { relationship: "Dad", gender: "Male" },
  "my dad": { relationship: "Dad", gender: "Male" },
  "my father": { relationship: "Dad", gender: "Male" },
  grandma: { relationship: "Grandma", gender: "Female" },
  grandmother: { relationship: "Grandma", gender: "Female" },
  grandpa: { relationship: "Grandpa", gender: "Male" },
  grandfather: { relationship: "Grandpa", gender: "Male" },
  wife: { relationship: "Spouse", gender: "Female" },
  "my wife": { relationship: "Spouse", gender: "Female" },
  husband: { relationship: "Spouse", gender: "Male" },
  "my husband": { relationship: "Spouse", gender: "Male" },
  spouse: { relationship: "Spouse", gender: "" },
  "my spouse": { relationship: "Spouse", gender: "" },
  myself: { relationship: "Myself", gender: "" },
  self: { relationship: "Myself", gender: "" },
  couple: { relationship: "Couple", gender: "" },
  relative: { relationship: "Relative", gender: "" },
  friend: { relationship: "Friend", gender: "" },
};

export function canonicalIdentityFacts(relationship: string): IdentityPatch {
  const trimmed = relationship.trim();
  return RELATIONSHIP_FACTS[trimmed.toLowerCase()] || { relationship: trimmed, gender: "" };
}

export function applyCanonicalIdentity<T extends IdentityState>(
  state: T,
  relationship: string,
  explicitGender = "",
): T {
  const facts = canonicalIdentityFacts(relationship);
  return {
    ...state,
    relationship: facts.relationship,
    gender: state.gender.trim() || explicitGender.trim() || facts.gender,
  } as T;
}

File: backend/app/services/canonical_intake_state.py
from __future__ import annotations

"""Canonicalize explicit intake identity facts before any decision layer reads them.

This module normalizes vocabulary; it does not infer a household, care need, or
relationship from narrative text.  Structured facts remain authoritative and an
explicit gender is never overwritten by a relationship-derived value.
"""

from copy import deepcopy
from typing import Any, Dict, Mapping


_RELATIONSHIP_ALIASES = {
    "mom": ("Mom", "Female"),
    "mother": ("Mom", "Female"),
    "my mom": ("Mom", "Female"),
    "my mother": ("Mom", "Female"),
    "dad": ("Dad", "Male"),
    "father": ("Dad", "Male"),
    "my dad": ("Dad", "Male"),
    "my father": ("Dad", "Male"),
    "grandma": ("Grandma", "Female"),
    "grandmother": ("Grandma", "Female"),
    "grandpa": ("Grandpa", "Male"),
    "grandfather": ("Grandpa", "Male"),
    "wife": ("Spouse", "Female"),
    "my wife": ("Spouse", "Female"),
    "husband": ("Spouse", "Male"),
    "my husband": ("Spouse", "Male"),
    "spouse": ("Spouse", ""),
    "my spouse": ("Spouse", ""),
    "myself": ("Myself", ""),
    "self": ("Myself", ""),
    "couple": ("Couple", ""),
    "relative": ("Relative", ""),
    "friend": ("Friend", ""),
}

_GENDER_ALIASES = {
    "male": "Male",
    "man": "Male",
    "female": "Female",
    "woman": "Female",
    "nonbinary": "Nonbinary",
    "non-binary": "Nonbinary",
    "other": "Other",
    "prefer not to say": "Prefer not to say",
}


def canonicalize_intake_state(questionnaire_state: Mapping[str, Any] | None) -> Dict[str, Any]:
    """Return an idempotent copy with explicit identity vocabulary normalized."""

    state: Dict[str, Any] = deepcopy(dict(questionnaire_state or {}))
    raw_relationship = str(state.get("relationship") or "").strip()
    raw_gender = str(state.get("gender") or "").strip()

    relationship, relationship_gender = _RELATIONSHIP_ALIASES.get(
        raw_relationship.casefold(),
        (raw_relationship, ""),
    )
    if relationship:
        state["relationship"] = relationship

    canonical_gender = _GENDER_ALIASES.get(raw_gender.casefold(), raw_gender)
    if canonical_gender:
        state["gender"] = canonical_gender
    elif relationship_gender:
        state["gender"] = relationship_gender

    return state
